File: Algorithm/c_or_cpp/charpter9_binary/level2/search_BST.cpp
#include <vector>
#include <stack>
#include <iostream>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// 迭代法
TreeNode *searchBST1(TreeNode *root, int val) {
        while (root) {
            if (val == root->val) {
                return root;
            }
            root = val < root->val ? root->left : root->right;
        }
        return nullptr;
    }

 //递归法
TreeNode *searchBST(TreeNode *root, int val)
{
    if (root == nullptr)
    {
        return nullptr;
    }
    if (val == root->val)
    {
        return root;
    }
    return searchBST(val < root->val ? root->left : root->right, val);
}

int main()
{
    // 创建测试用例的二叉树
    TreeNode *root1 = new TreeNode(5);
    root1->left = new TreeNode(3);
    root1->right = new TreeNode(7);
    root1->left->left = new TreeNode(1);
    root1->left->right = new TreeNode(2);
    root1->right->left = new TreeNode(6);
    root1->right->right = new TreeNode(8);
    TreeNode *res = searchBST(root1, 5);

    return 0;
}

File: Algorithm/Code_Materials/Java/src/main/java/com/yugutou/charpter9_bitree_bisearch/level2/SearchBST.java
package com.yugutou.charpter9_bitree_bisearch.level2;


import com.yugutou.tools.BinaryTree;
import com.yugutou.tools.TreeNode;

/**
 * LeetCode 700.给定二叉搜索树（BST）的根节点和一个值。
 * 你需要在BST中找到节点值等于给定值的节点。 返回以该节点为根的子树。
 */
public class SearchBST {
    public static void main(String[] args) {
        BinaryTree bTree = new BinaryTree();
        bTree.root = bTree.buildBSTTree();
        System.out.println(searchBST(bTree.root, 2).val);
        System.out.println(searchBST2(bTree.root, 2).val);
    }

    /**
     * 递归方式实现
     *
     * @param root
     * @param val
     * @return
     */
    public static TreeNode searchBST(TreeNode root, int val) {
        if (root == null || val == root.val) return root;
        return val < root.val ? searchBST(root.left, val) : searchBST(root.right, val);
    }

    /**
     * 迭代实现
     *
     * @param root
     * @param val
     * @return
     */
    public static TreeNode searchBST2(TreeNode root, int val) {
        while (root != null && val != root.val)
            root = val < root.val ? root.left : root.right;
        return root;
    }
}
